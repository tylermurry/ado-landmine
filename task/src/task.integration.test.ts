import tl = require('azure-pipelines-task-lib/task');
import { WebApi } from 'azure-devops-node-api';
import {CommentThreadStatus} from "azure-devops-node-api/interfaces/GitInterfaces";
import { table } from 'table';
import executeTask from './task';

jest.mock('azure-devops-node-api');
jest.mock('azure-pipelines-task-lib/task');
jest.mock('table');

const gitApiMock = {
    getThreads: jest.fn(),
    createComment: jest.fn(),
    updateThread: jest.fn(),
};
(WebApi as any).mockImplementation(() => ({ getGitApi: () => gitApiMock }));

jest.setTimeout(10000000);

describe('Integration Test', () => {
    beforeEach(() => {
        jest.clearAllMocks();

        // Mock Inputs
        (tl.getInput as any).mockReturnValueOnce('some-token');
        (tl.getInput as any).mockReturnValueOnce('npm test');
        (tl.getInput as any).mockReturnValueOnce('./src/__mock__');
        (tl.getInput as any).mockReturnValueOnce('5000');
        (tl.getInput as any).mockReturnValueOnce(undefined);
        process.env.SYSTEM_TEAMPROJECT = 'team-project';
        process.env.BUILD_REPOSITORY_NAME = 'some-repo';
        process.env.SYSTEM_PULLREQUEST_PULLREQUESTID = '123456';
        process.env.SYSTEM_TEAMFOUNDATIONCOLLECTIONURI = 'some-org-url';
    })

    it('should add two landmines and catch them all', async () => {
        gitApiMock.getThreads.mockReturnValueOnce([{
            id: 1,
            status: CommentThreadStatus.Active,
            comments: [
                { content: "/bomb\n\n```suggestion\n2\n```" },
            ],
            threadContext: {
                filePath: '/insertion-sort.js',
                rightFileStart: {line: 3, offset: 18},
                rightFileEnd: {line: 3, offset: 19},
            }
        }, {
            id: 2,
            status: CommentThreadStatus.Active,
            comments: [
                { content: "ðŸ’£\n\n```suggestion\n3\n```" },
                { content: "ðŸ’¥ Bomb not defused. Verify your tests can catch this issue." },
            ],
            threadContext: {
                filePath: '/insertion-sort.js',
                rightFileStart: {line: 3, offset: 18},
                rightFileEnd: {line: 3, offset: 19},
            }
        }]);

        await executeTask();

        expect(tl.setResult).toMatchSnapshot();
        expect(table).toMatchSnapshot();
        expect(gitApiMock.createComment).toMatchSnapshot();
        expect(gitApiMock.updateThread).toMatchSnapshot();
    });

    it('should add a landmine but not catch it', async () => {
        gitApiMock.getThreads.mockReturnValueOnce([{
            id: 1,
            status: CommentThreadStatus.Active,
            comments: [
                { content: "/bomb\n\n```suggestion\n; console.log('here')\n```" },
            ],
            threadContext: {
                filePath: '/insertion-sort.js',
                rightFileStart: {line: 2, offset: 33},
                rightFileEnd: {line: 2, offset: 34},
            }
        }]);

        await executeTask();

        expect(tl.setResult).toMatchSnapshot();
        expect(table).toMatchSnapshot();
        expect(gitApiMock.createComment).toMatchSnapshot();
        expect(gitApiMock.updateThread).toMatchSnapshot();
    });

    it('should add a landmine but not catch it because the execution timed out', async () => {
        gitApiMock.getThreads.mockReturnValueOnce([{
            id: 1,
            status: CommentThreadStatus.Active,
            comments: [
                { content: "ðŸ’£\n\n```suggestion\nconsole.log('infinite loop!');\n```" },
            ],
            threadContext: {
                filePath: '/insertion-sort.js',
                rightFileStart: {line: 7, offset: 13},
                rightFileEnd: {line: 8, offset: 23},
            }
        }]);

        await executeTask();

        expect(tl.setResult).toMatchSnapshot();
        expect(table).toMatchSnapshot();
        expect(gitApiMock.createComment).toMatchSnapshot();
        expect(gitApiMock.updateThread).toMatchSnapshot();
    });

    it('should not find any landmines because there are no threads', async () => {
        gitApiMock.getThreads.mockReturnValueOnce([]);

        await executeTask();

        expect(tl.setResult).toMatchSnapshot();
        expect(table).not.toHaveBeenCalled();
        expect(gitApiMock.createComment).toMatchSnapshot();
        expect(gitApiMock.updateThread).toMatchSnapshot();
    });

    it('should not find any landmines because there are no active threads', async () => {
        gitApiMock.getThreads.mockReturnValueOnce([{
            id: 1,
            status: CommentThreadStatus.Active,
            comments: [
                { content: "ðŸ’£\n\n```suggestion\n2\n```" },
                { content: "âœ… Bomb successfully defused" },
            ],
            threadContext: {
                filePath: '/insertion-sort.js',
                rightFileStart: {line: 3, offset: 18},
                rightFileEnd: {line: 3, offset: 19},
            }
        },{
            id: 2,
            status: CommentThreadStatus.Closed,
            comments: [
                { content: "ðŸ’£\n\n```suggestion\n3\n```" },
            ],
            threadContext: {
                filePath: '/insertion-sort.js',
                rightFileStart: {line: 3, offset: 18},
                rightFileEnd: {line: 3, offset: 19},
            }
        }]);

        await executeTask();

        expect(tl.setResult).toMatchSnapshot();
        expect(table).not.toHaveBeenCalled();
        expect(gitApiMock.createComment).toMatchSnapshot();
        expect(gitApiMock.updateThread).toMatchSnapshot();
    });
});
