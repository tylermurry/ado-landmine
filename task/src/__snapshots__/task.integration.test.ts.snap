// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration Test should add a landmine but not catch it 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      2,
      "There was at least bomb that was not defused.",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Integration Test should add a landmine but not catch it 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Array [
        Array [
          "File Name",
          "Line Number",
          "Bomb Defused",
        ],
        Array [
          "insertion-sort.js",
          2,
          "no",
        ],
      ],
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Integration Test should add a landmine but not catch it 3`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "content": "ðŸ’¥ Bomb not defused. Please adjust your test to catch the error",
      },
      "some-repo",
      123456,
      1,
      "team-project",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Integration Test should add a landmine but not catch it 4`] = `[MockFunction]`;

exports[`Integration Test should add a landmine but not catch it because the execution timed out 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      2,
      "There was at least bomb that was not defused.",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Integration Test should add a landmine but not catch it because the execution timed out 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Array [
        Array [
          "File Name",
          "Line Number",
          "Bomb Defused",
        ],
        Array [
          "insertion-sort.js",
          "7-8",
          "no",
        ],
      ],
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Integration Test should add a landmine but not catch it because the execution timed out 3`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "content": "ðŸ’¥ Bomb not defused. Please adjust your test to catch the error",
      },
      "some-repo",
      123456,
      1,
      "team-project",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Integration Test should add a landmine but not catch it because the execution timed out 4`] = `[MockFunction]`;

exports[`Integration Test should add two landmines and catch them all 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      0,
      "All bombs successfully defused",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Integration Test should add two landmines and catch them all 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Array [
        Array [
          "File Name",
          "Line Number",
          "Bomb Defused",
        ],
        Array [
          "insertion-sort.js",
          3,
          "yes",
        ],
        Array [
          "insertion-sort.js",
          3,
          "yes",
        ],
      ],
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Integration Test should add two landmines and catch them all 3`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "content": "âœ… Successfully defused bomb",
      },
      "some-repo",
      123456,
      1,
      "team-project",
    ],
    Array [
      Object {
        "content": "âœ… Successfully defused bomb",
      },
      "some-repo",
      123456,
      2,
      "team-project",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Integration Test should add two landmines and catch them all 4`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "status": 2,
      },
      "some-repo",
      123456,
      1,
      "team-project",
    ],
    Array [
      Object {
        "status": 2,
      },
      "some-repo",
      123456,
      2,
      "team-project",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Integration Test should not find any landmines because there are no active threads 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      0,
      "All bombs successfully defused",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Integration Test should not find any landmines because there are no active threads 2`] = `[MockFunction]`;

exports[`Integration Test should not find any landmines because there are no active threads 3`] = `[MockFunction]`;

exports[`Integration Test should not find any landmines because there are no active threads 4`] = `[MockFunction]`;

exports[`Integration Test should not find any landmines because there are no threads 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      0,
      "All bombs successfully defused",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Integration Test should not find any landmines because there are no threads 2`] = `[MockFunction]`;

exports[`Integration Test should not find any landmines because there are no threads 3`] = `[MockFunction]`;

exports[`Integration Test should not find any landmines because there are no threads 4`] = `[MockFunction]`;
